src/coreclr/nativeaot/System.Private.CoreLib/src/System/Diagnostics/StackTrace.NativeAot.Browser.cs

```diff
@@ -1,65 +1,93 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Text;
+using System.Runtime;
+using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
 namespace System.Diagnostics
 {
     public partial class StackTrace
     {
-        private readonly StringBuilder _builder = new StringBuilder();
+        /// <summary>
+        /// Initialize the stack trace based on current thread and given initial frame index.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private unsafe void InitializeForCurrentThread(int skipFrames, bool needFileInfo)
+        {
+            const int SystemDiagnosticsStackDepth = 2;
 
-        [LibraryImport("*")]
-        private static unsafe partial int emscripten_get_callstack(int flags, byte* outBuf, int maxBytes);
+            int eipCount = RuntimeImports.RhpGetCurrentBrowserThreadStackTrace(0, Exception.ReportAllFramesAsJS());
+            IntPtr[] eips = new IntPtr[eipCount];
+            fixed (void* pEips = eips)
+                RuntimeImports.RhpGetCurrentBrowserThreadStackTrace((nuint)pEips, Exception.ReportAllFramesAsJS());
 
-        private unsafe void InitializeForCurrentThread(int skipFrames, bool needFileInfo)
+            int skippedEipCount = Exception.SkipSystemFrames(eips, SystemDiagnosticsStackDepth);
+            InitializeForIpAddressArray(eips, skippedEipCount, skipFrames, needFileInfo);
+        }
+
+        /// <summary>
+        /// Initialize the stack trace based on a given exception and initial frame index.
+        /// </summary>
+        private void InitializeForException(Exception exception, int skipFrames, bool needFileInfo)
+        {
+            InitializeForIpAddressArray(exception.GetStackIPs(), 0, skipFrames, needFileInfo);
+        }
+
+        /// <summary>
+        /// Initialize the stack trace based on a given array of encoded IP addresses (EIPs).
+        /// </summary>
+        private void InitializeForIpAddressArray(IntPtr[] eips, int skippedEipCount, int skipFrames, bool needFileInfo)
         {
-            var backtraceBuffer = new byte[8192];
-            int callstackLen;
-            // skip these 2:
-            // at S_P_CoreLib_System_Diagnostics_StackTrace__InitializeForCurrentThread (wasm-function[12314]:275)
-            // at S_P_CoreLib_System_Diagnostics_StackTrace___ctor_0(wasm-function[12724]:118)
-            skipFrames += 2; // METHODS_TO_SKIP is a constant so just change here
+            // Our callers may pass us values that have overflown.
```

Comment:
It turns out that it is legal to call `new StackTrace(skipFrames: int.MinValue)` and such, and there are even tests verifying the behavior of that. Such values are probably the result of overflow.

Also, the initial version of this code just did `skipFrames += SystemDiagnosticsStackDepth;` (like in `StackFrame.NativeAot.Browser.cs`) instead of `SkipSystemFrames`, which would overflow for `new StackTrace(skipFrames: int.MaxValue)`.

(Q: should the stack frame logic should also use `SkipSystemFrames`? A: probably not, since it would mean that in case of JSC or modified binaries it would start yielding system frames only. Which is not too bad for `new StackTrace().ToString()`, but makes `new StackFrame()` basically useless.)